packages/server/modules/core/migrations/20240710154658_user_emails_backfill.ts

```diff
@@ -0,0 +1,35 @@
+import { UserRecord } from '@/modules/core/helpers/types'
+import { Knex } from 'knex'
+import crs from 'crypto-random-string'
+
+export async function up(knex: Knex): Promise<void> {
+  const batchSize = 1000
+  const offset = 0
+  let users = []
+
+  do {
+    users = await knex<UserRecord>('users')
+      .select(['id', 'email', 'verified'])
+      .limit(batchSize)
+      .offset(offset)
+
+    if (users.length === 0) return
+
+    await knex('user_emails')
+      .insert(
+        users.map((user) => ({
+          id: crs({ length: 10 }),
+          userId: user.id,
+          email: user.email,
+          verified: user.verified,
+          primary: true
+        }))
+      )
+      .onConflict(['userId', 'email'])
+      .ignore()
+  } while (users.length > 0)
```

Comment:
Always add failsafes for while/do-while loops, cause if you have a bug in there, you could run into an infinite loop. It can be an impractical limit that you would never actually reach, but it would at least ensure it stops somewhere - e.g. after a 100 pages.

This specific migration actually would've caused exactly that - you never increase the offset, so when this would run in prod, it would inifinitely run the same queries on the same set of first 1000 users